src/dft/backends/cufft/forward.cpp

```diff
@@ -0,0 +1,179 @@
+/*******************************************************************************
+* Copyright Codeplay Software Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#include <type_traits>
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include "oneapi/mkl/types.hpp"
+#include "oneapi/mkl/exceptions.hpp"
+
+#include "oneapi/mkl/dft/detail/cufft/onemkl_dft_cufft.hpp"
+#include "oneapi/mkl/dft/detail/types_impl.hpp"
+#include "oneapi/mkl/dft/detail/descriptor_impl.hpp"
+
+// cuFFT headers
+#include <cuda_runtime.h>
+#include <cufft.h>
+
+namespace oneapi::mkl::dft::cufft {
+namespace detail {
+
+template <dft::detail::precision prec, dft::detail::domain dom>
+inline dft::detail::commit_impl *get_commit(dft::detail::descriptor<prec, dom> &desc) {
+    auto commit_handle = dft::detail::get_commit(desc);
+    if (commit_handle == nullptr || commit_handle->get_backend() != backend::cufft) {
+        throw mkl::invalid_argument("DFT", "compute_backward",
+                                    "DFT descriptor has not been commited for cuFFT");
+    }
+    return commit_handle;
+}
+
+/// Throw an mkl::invalid_argument if the runtime param in the descriptor does not match
+/// the expected value.
+template <dft::detail::config_param Param, dft::detail::config_value Expected, typename DescT>
+inline auto expect_config(DescT &desc, const char *message) {
+    dft::detail::config_value actual{ 0 };
+    desc.get_value(Param, &actual);
+    if (actual != Expected) {
+        throw mkl::invalid_argument("DFT", "compute_forward", message);
+    }
+}
+} // namespace detail
```

Comment:
Are these two functions copies of the ones in backward.hpp?

Reply:
They should have had different arguments for the throw, that is fixed in [8907ca38b](https://github.com/oneapi-src/oneMKL/pull/278/commits/8907ca38b4c8cacfc3d9fe5d5e6cd434e4f2f61b).